### PR TITLE
[Fluid] Remove statistics app dependency in python

### PR DESCRIPTION
**Pull request by sunethwarna**
**📝 Description**
This PR is related to concerns raised by @rubenzorrilla in PR #10097

This PR removes currently existing `StatisticsApplication` dependency in `FluidDynamicsApplication`. It introduces `compute_cfl_process` to calculate currently only the `CFL_NUMBER` number and `compute_y_plus_process` to calculate only the `Y_PLUS`. Previously with the "cfl_output_process.py", cfl output was written to a file. This can now be achieved by using `compute_cfl_process` and `SpatialStatisticsProcess` in `StatisticsApplication`. This also removes `y_plus_output_process.py`. Previously obtained output from both `cfl_output_process.py` and `y_plus_output_process.py` now can be obtained by using `compute_cfl_process`, `compute_y_plus_process` and `SpatialStatisticsProcess` from `StatisticsApplication`. Here are some examples:

This PR depends on #10182.

Old way of calculating y_plus_output:
```json
[
   {
            "kratos_module"   : "KratosMultiphysics.FluidDynamicsApplication",
            "python_module"   : "cfl_output_process",
            "help"            : "",
            "Parameters" : {
                "model_part_name" : "MainModelPart.Parts_Fluid",
                "interval"             : [0.0, 1e30],
                "cfl_output_limit"     : 2.5,
                "print_to_screen"      : false,
                "print_format"         : ".8f",
                "write_output_file"    : true,
                "output_step"          : 2,
                "output_file_settings": {
                    "file_name": "cfl_output_process.dat"
                }
            }
   }
]
```
New way of calculating it:
```json
[
            {
                "kratos_module": "KratosMultiphysics.StatisticsApplication",
                "python_module": "spatial_statistics_process",
                "Parameters": {
                    "model_part_name": "MainModelPart.Parts_Fluid",
                    "computation_processes": [
                        {
                            "kratos_module": "KratosMultiphysics.FluidDynamicsApplication",
                            "python_module": "compute_cfl_process",
                            "Parameters": {
                                "model_part_name": "MainModelPart.Parts_Fluid",
                                "echo_level": 0
                            }
                        }
                    ],
                    "input_variable_settings": [
                        {
                            "norm_type": "magnitude",
                            "container": "element_non_historical",
                            "variable_names": [
                                "CFL_NUMBER"
                            ]
                        }
                    ],
                    "statistics_methods": [
                        {
                            "method_name": "variance"
                        },
                        {
                            "method_name": "max"
                        },
                        {
                            "method_name": "distribution",
                            "method_settings": {
                                "number_of_value_groups": 1,
                                "min_value": 1.0,
                                "max_value": 2.5
                            }
                        }
                    ],
                    "output_settings": {
                        "output_control_variable": "STEP",
                        "output_time_interval": 1,
                        "write_kratos_version": true,
                        "write_time_stamp": true,
                        "output_value_precision": 2,
                        "output_value_length": 6,
                        "output_file_settings": {
                            "file_name": "<model_part_name>_cfl.dat",
                            "output_path": "spatial_statistics_process",
                            "write_buffer_size": -1
                        }
                    }
                }
            }
]
```
The output of CFL number will look like the following
```
# Spatial statistics process output
# Kratos version               : 9.1."3"-1396aa0a9b-Release-x86_64
# Timestamp                    : 2022-08-23 09:16:20.836518
# Method name(s)               : variance, max, distribution
# Variable info(s)             : CFL_NUMBER[magnitude, element_non_historical]
# Modelpart name               : MainModelPart.Parts_Fluid
# Output control variable name : STEP
# ----------------------------------------------------------------------
# Headers:
#"TIME" "STEP" "mean" "variance" "max_value" "max_id"                    "group_below_min {mean|variance}"                            "group_1 {mean|variance}"                    "group_above_max {mean|variance}"
   0.20      2   1.98       1.36        4.32       50     13 [#] =  26.00 [%] <   1.00 {   0.54 |   0.08 }     20 [#] =  40.00 [%] <   2.50 {   1.79 |   0.20 }     17 [#] =  34.00 [%]>=   2.50 {   3.31 |   0.32 }
   0.40      4   3.95       5.43        8.62       50      6 [#] =  12.00 [%] <   1.00 {   0.57 |   0.08 }      9 [#] =  18.00 [%] <   2.50 {   1.69 |   0.19 }     35 [#] =  70.00 [%]>=   2.50 {   5.12 |   3.06 }
   0.60      6   5.93      12.22       12.92       50      3 [#] =   6.00 [%] <   1.00 {   0.50 |   0.05 }      7 [#] =  14.00 [%] <   2.50 {   1.62 |   0.22 }     40 [#] =  80.00 [%]>=   2.50 {   7.09 |   8.44 }
   0.80      8   7.90      21.75       17.24       50      2 [#] =   4.00 [%] <   1.00 {   0.45 |   0.00 }      6 [#] =  12.00 [%] <   2.50 {   1.70 |   0.17 }     42 [#] =  84.00 [%]>=   2.50 {   9.14 |  16.20 }
   1.00     10   9.87      34.04       21.57       50      2 [#] =   4.00 [%] <   1.00 {   0.57 |   0.00 }      4 [#] =   8.00 [%] <   2.50 {   1.83 |   0.15 }     44 [#] =  88.00 [%]>=   2.50 {  11.02 |  27.54 }
# End of file.
```

The y+ with its statistics can be computed using following json settings. Make sure to calculate reactions from the solver since `compute_y_plus_process` uses nodal `REACTIONS` to compute `y+`.
```json
            {
                "kratos_module": "KratosMultiphysics.StatisticsApplication",
                "python_module": "spatial_statistics_process",
                "Parameters": {
                    "model_part_name": "MainModelPart",
                    "computation_processes": [
                        {
                            "kratos_module": "KratosMultiphysics.FluidDynamicsApplication",
                            "python_module": "compute_y_plus_process",
                            "Parameters": {
                                "model_part_name": "MainModelPart",
                                "echo_level": 2
                            }
                        }
                    ],
                    "input_variable_settings": [
                        {
                            "norm_type": "magnitude",
                            "container": "condition_non_historical",
                            "variable_names": [
                                "Y_PLUS"
                            ]
                        }
                    ],
                    "statistics_methods": [
                        {
                            "method_name": "variance"
                        },
                        {
                            "method_name": "max"
                        },
                        {
                            "method_name": "distribution",
                            "method_settings": {
                                "number_of_value_groups": 1,
                                "min_value": 50.0,
                                "max_value": 300.0
                            }
                        }
                    ],
                    "output_settings": {
                        "output_control_variable": "STEP",
                        "output_time_interval": 1,
                        "write_kratos_version": true,
                        "write_time_stamp": true,
                        "output_value_precision": 2,
                        "output_value_length": 6,
                        "output_file_settings": {
                            "file_name": "<model_part_name>_y_plus.dat",
                            "output_path": "spatial_statistics_process",
                            "write_buffer_size": -1
                        }
                    }
                }
            }
```

The `y+` output looks like the following:
```
# Spatial statistics process output
# Kratos version               : 9.1."3"-3a06cd7c4a-Release-x86_64
# Timestamp                    : 2022-08-23 09:44:43.804304
# Method name(s)               : variance, max, distribution
# Variable info(s)             : Y_PLUS[magnitude, condition_non_historical]
# Modelpart name               : MainModelPart
# Output control variable name : STEP
# ----------------------------------------------------------------------
# Headers:
#  "TIME"   "STEP"   "mean" "variance" "max_value" "max_id"                              "group_below_min {mean|variance}"                                      "group_1 {mean|variance}"                              "group_above_max {mean|variance}"
     0.20        2     2.48       1.79        4.17       10       20 [#] =   100.00 [%] <    50.00 {     2.48 |     1.79 }        0 [#] =     0.00 [%] <   300.00 {     0.00 |     0.00 }        0 [#] =     0.00 [%]>=   300.00 {     0.00 |     0.00 }
     0.40        4     3.51       3.57        5.90       10       20 [#] =   100.00 [%] <    50.00 {     3.51 |     3.57 }        0 [#] =     0.00 [%] <   300.00 {     0.00 |     0.00 }        0 [#] =     0.00 [%]>=   300.00 {     0.00 |     0.00 }
     0.60        6     4.30       5.36        7.22       10       20 [#] =   100.00 [%] <    50.00 {     4.30 |     5.36 }        0 [#] =     0.00 [%] <   300.00 {     0.00 |     0.00 }        0 [#] =     0.00 [%]>=   300.00 {     0.00 |     0.00 }
     0.80        8     4.97       7.14        8.34       10       20 [#] =   100.00 [%] <    50.00 {     4.97 |     7.14 }        0 [#] =     0.00 [%] <   300.00 {     0.00 |     0.00 }        0 [#] =     0.00 [%]>=   300.00 {     0.00 |     0.00 }
     1.00       10     5.55       8.93        9.33       10       20 [#] =   100.00 [%] <    50.00 {     5.55 |     8.93 }        0 [#] =     0.00 [%] <   300.00 {     0.00 |     0.00 }        0 [#] =     0.00 [%]>=   300.00 {     0.00 |     0.00 }
# End of file.
```

The output of the `SpatialStatisticsProcess` is explained in https://kratosmultiphysics.github.io/Kratos/pages/Applications/Statistics_Application/Examples/spatial_example.html

**🆕 Changelog**
- Removed `CFLOutputProcess`
- Removed `YPlusOutputProcess`
- Removed `CFLOutputProcess` test
- Added `compute_cfl_process`
- Added `compute_y_plus_process`
- Added tests

### CODE REVIEW

**Review by rubenzorrilla — CHANGES_REQUESTED** [review]
I'm a bit hesitant about this design... Here my comments about it

1. I think that its usage is complex. Note that for doing one operation (e.g. compute some statistics of the CFL number) the user needs to add two processes to the json settings.
2. As far as I see, there is no guarantee that the magnitude of interest is updated when the statistics are computed as it relies on the sorting of the json list.
3. `FluidUtilitiesProcess` calculates the magnitudes at each step, even though these are not used by the statistics.

Here it comes my suggestion. I'd extend the `spatial_statistics_process` (and I assume that we'll need to do so in the `temporal_statistics_process`) to accept an extra `auxiliary_operation_module` and `auxiliary_-operation_settings` fields. These will contain the Python module that implements an operation (note that an operation is a class which public API only features an `Execute` method to be called and its corresponding settings. By default these are empty, so we keep current behavior of the statistics process by default. If these are set, we instantiate the class in `auxiliary_operation_module` with the `auxilary_operation_settings` dict. If this is the case, we call the `Execute` method from within the statistics process each time is required. 

Nice thing of this design is that the user will only set the statistics process with all the related information. Besides such variables are calculated only if required.

I've also other minor comments regarding the `fluid_utilities_process` but I'd only post them once we figure out the final design.

**Comment by mpentek** [comment]
I am also hesitant about the new design. As far as user-experience, the previous one is much more straightforward. I understand that the new was is thought to be modular and generic, hence better coding practice.

What about the following: the user can define it in the old way or the new way? The old way acts simply as an "input mask" and creates the blocks for the new syntax.

**Comment by sunethwarna** [comment]
@mpentek I think the most nicer way is to go along with what @rubenzorrilla is suggesting, creating a process in statistics which can include a utility process which has `Execute` method defined. Having a masking mechanism will still need the `StatisticsAppliation` in fluid app, if it to be put on `StatisticsApplication` then it will have dependency in `FluidApp` therefore unfortunately another complicated dependency will rise. I will work on the way @rubenzorrilla  has suggested and ping you all again.

**Comment by rubenzorrilla** [comment]
> @mpentek I think the most nicer way is to go along with what @rubenzorrilla is suggesting, creating a process in statistics which can include a utility process which has `Execute` method defined. Having a masking mechanism will still need the `StatisticsAppliation` in fluid app, if it to be put on `StatisticsApplication` then it will have dependency in `FluidApp` therefore unfortunately another complicated dependency will rise. I will work on the way @rubenzorrilla has suggested and ping you all again.

Very good. I think that my proposal sounds more complex than what it is. Besides it could be easily extended to other similar use cases.

**Comment by sunethwarna** [comment]
@rubenzorrilla I have updated the branch and included some example json blocks to illustrate how it can be used to obtain statistics. It is little bit verbose than the settings what we used to have. But I unfortunately cannot avoid it if we are to remove application dependencies.

fyi @mpentek @mtnzguillermo

**Comment by rubenzorrilla** [comment]
> @rubenzorrilla I have updated the branch and included some example json blocks to illustrate how it can be used to obtain statistics. It is little bit verbose than the settings what we used to have. But I unfortunately cannot avoid it if we are to remove application dependencies.
> 
> fyi @mpentek @mtnzguillermo

Well, in a standard situation (with no processes) it's as verbose as before. The more auxiliary processes you use the more verbose it gets, but this is what it is so I don't think we should be worried about verbosity in this case. Nice thing is that everything is encapsulated in one place (so much easy to use) and that it is super flexible. I like it a lot. Good job 🎉

File: applications/FluidDynamicsApplication/tests/test_fluid_computation_processes.py [hunk 1]
```diff
@@ -0,0 +1,82 @@
+import KratosMultiphysics as Kratos
+import KratosMultiphysics.KratosUnittest as UnitTest
+from KratosMultiphysics.testing.utilities import ReadModelPart
+from KratosMultiphysics.process_factory import KratosProcessFactory
+
+class FluidComputationProcessTest(UnitTest.TestCase):
```
Comment (rubenzorrilla) [inline applications/FluidDynamicsApplication/tests/test_fluid_computation_processes.py]:
```suggestion
class FluidComputationProcessesTest(UnitTest.TestCase):
```
To be consistent with the file name and make clear that we will test similar auxiliary Python modules in here.

File: applications/FluidDynamicsApplication/tests/test_FluidDynamicsApplication.py [hunk 2]
```diff
@@ -52,7 +51,7 @@
 from compute_pressure_coefficient_process_test import ComputePressureCoefficientProcessTest
 from compute_drag_process_test import ComputeDragProcessTest
 from test_compute_y_plus_process import ComputeYPlusProcessTest
-
+from test_fluid_computation_processes import FluidComputationProcessTest
```
Comment (rubenzorrilla) [inline applications/FluidDynamicsApplication/tests/test_FluidDynamicsApplication.py]:
```suggestion
from test_fluid_computation_processes import FluidComputationProcessesTest
```

File: applications/FluidDynamicsApplication/tests/test_FluidDynamicsApplication.py [hunk 3]
```diff
@@ -127,11 +126,11 @@ def AssembleTestSuites():
     nightSuite.addTest(FluidAnalysisTest('testSteadyCylinder'))
     nightSuite.addTest(ConsistentLevelsetNodalGradientTest('testConsistentGradientSquare2D'))
     nightSuite.addTests(KratosUnittest.TestLoader().loadTestsFromTestCases([IntegrationPointStatisticsTest]))
-    nightSuite.addTests(KratosUnittest.TestLoader().loadTestsFromTestCases([CFLOutputProcessTest]))
     nightSuite.addTests(KratosUnittest.TestLoader().loadTestsFromTestCases([FlowsMeasuringUtilityTest]))
     nightSuite.addTests(KratosUnittest.TestLoader().loadTestsFromTestCases([FluidAuxiliaryUtilitiesTest]))
     nightSuite.addTests(KratosUnittest.TestLoader().loadTestsFromTestCases([TwoFluidMassConservationTest]))
     nightSuite.addTests(KratosUnittest.TestLoader().loadTestsFromTestCases([NavierStokesCompressibleExplicitSolverTest]))
+    nightSuite.addTests(KratosUnittest.TestLoader().loadTestsFromTestCases([FluidComputationProcessTest]))
```
Comment (rubenzorrilla) [inline applications/FluidDynamicsApplication/tests/test_FluidDynamicsApplication.py]:
```suggestion
    nightSuite.addTests(KratosUnittest.TestLoader().loadTestsFromTestCases([FluidComputationProcessesTest]))
```

File: applications/FluidDynamicsApplication/tests/test_fluid_computation_processes.py [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+        for step, time_step in enumerate(time_steps, 1):
+            self.model_part.ProcessInfo[Kratos.TIME] = time_step
+            self.model_part.ProcessInfo[Kratos.STEP] = step
+
+            for process in process_list:
+                process.ExecuteInitializeSolutionStep()
+                process.ExecuteFinalizeSolutionStep()
+
+        for process in process_list:
+            process.ExecuteFinalize()
```
Comment (rubenzorrilla) [inline applications/FluidDynamicsApplication/tests/test_fluid_computation_processes.py]: Shouldn't you call the `Execute` in here?

**Review by rubenzorrilla — COMMENTED** [review]
I'm very positive about the changes. However I don't fully understand the testing. In particular, I don't see where are and which are the reference values you are testing to.

File: applications/FluidDynamicsApplication/tests/test_fluid_computation_processes.py (hunk 4 above)
Comment (sunethwarna) [reply applications/FluidDynamicsApplication/tests/test_fluid_computation_processes.py]: Ooops... I forgot.

**Comment by sunethwarna** [comment]
Testing was done to check whether the processes work smoothly only. There are separate tests where I check the values such as 
https://github.com/KratosMultiphysics/Kratos/blob/master/applications/FluidDynamicsApplication/tests/test_compute_y_plus_process.py

**Comment by rubenzorrilla** [comment]
> Testing was done to check whether the processes work smoothly only. There are separate tests where I check the values such as https://github.com/KratosMultiphysics/Kratos/blob/master/applications/FluidDynamicsApplication/tests/test_compute_y_plus_process.py

Fair enough. In any case I'd put a short comment explaining this in the test. :+1:

**Review by rubenzorrilla — APPROVED** [review]

**Pull request merged**